File: yoojongho/2024-06-30-rotate-list.java
/**
 * Leetcode
 * problem: 61
 * link: https://leetcode.com/problems/rotate-list/
 * tag: Linked List, Two Pointers
 */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) {
            return head;
        }

        int length = 1;
        ListNode temp = head;

        while (temp.next != null) {
            temp = temp.next;
            length++;
        }
        temp.next = head;
        k = k % length;
        k = length - k;

        while (k-- > 0) {
            temp = temp.next;
        }
        // 123451234512345..., head.val = 1, temp.val = 3
        head = temp.next;
        // 451234512345123..., head.val = 4, temp.val = 3
        temp.next = null;
        // 45213
        return head;
    }
}

File: Changwhi/2024-06-30-Rotate-List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        lastNode = head
        count = 1
        # You should use "is not" or "is" when you compare to None
        # None is a singleton in Python, there is only one instance of it in memory
        # By comparing identity, this can be performed very quickly. 
        # Python checks whether the object you're referring to has the same memory address 
        # as the global None object - a very, very fast comparison of two numbers.
        while lastNode.next is not None:
            temp = lastNode
            lastNode = lastNode.next
            lastNode.prev = temp
            count += 1
        lastNode.next = head
        head.prev = lastNode
        
        if k > count:
            k = k % count
            
        currentNode = head
        for i in range(0, k + 1):
            currentNode = currentNode.prev
        head = currentNode.next
        currentNode.next = None
        
        return head        
            
        
            
        